[REVIEW] Add meta inference for `ParallelPostFit.predict` for sparse arrays

VibhuJawa opened the pull request:
This PR resolves https://github.com/dask/dask-ml/issues/888

Todo:
- [x] Add test for warning
- [x] Add test for ParallelPostFit

GenevieveBuckley commented on dask_ml/wrappers.py:
```diff
@@ -662,9 +663,28 @@ def _get_output_dask_ar_meta_for_estimator(model_fn, estimator, input_dask_ar):
+    elif "scipy.sparse" in type(imput_meta).__module__:
+        # sparse matrices dont support
+        # `like` due to non implimented __array_function__
+        # Refer https:/q/github.com/scipy/scipy/issues/10362
```
```suggestion
        # Refer https://github.com/scipy/scipy/issues/10362
```

Broken link is because of a typo here

GenevieveBuckley commented on dask_ml/wrappers.py:
```diff
@@ -662,9 +663,28 @@ def _get_output_dask_ar_meta_for_estimator(model_fn, estimator, input_dask_ar):
-        dtype=input_dask_ar.dtype,
-        like=input_dask_ar._meta,
-    )
+    imput_meta = input_dask_ar._meta
```
```suggestion
    input_meta = input_dask_ar._meta
```

Is this a typo? The variable name probably makes more sense if it says "input" (instead of "imput")

VibhuJawa replied:
Yup. That was a typo. Thanks for the review. Pushed a fix with commit . https://github.com/dask/dask-ml/pull/889/commits/a295e39d11826c60faf43e0d1d42dc66a9a0a265

GenevieveBuckley commented on dask_ml/wrappers.py:
```diff
@@ -662,9 +663,28 @@ def _get_output_dask_ar_meta_for_estimator(model_fn, estimator, input_dask_ar):
+        # Refer https:/q/github.com/scipy/scipy/issues/10362
+        # Note below works for both cupy and scipy sparse matrices
+        ar = type(imput_meta)((1, input_dask_ar.shape[1]), dtype=input_dask_ar.dtype)
+    else:
```
When does this clause get triggered? Could we have a test for it?

GenevieveBuckley replied:
Ah, I see from your todo list you're planning on adding a test, that's great!

VibhuJawa replied:
Yup. Added a test [here](https://github.com/dask/dask-ml/blob/48645448147126c6a343d6d085dd571be0cd1665/tests/test_parallel_post_fit.py#L253-L291) . 

This basically gets triggered when using dask array wrapper around an object that does not support `__array_function__` .   A real world common example will be a Tensorflow tensor

GenevieveBuckley commented on dask_ml/wrappers.py:
```diff
@@ -662,9 +663,28 @@ def _get_output_dask_ar_meta_for_estimator(model_fn, estimator, input_dask_ar):
+        ar = type(imput_meta)((1, input_dask_ar.shape[1]), dtype=input_dask_ar.dtype)
+    else:
+        msg = (
+            "\nYou did not provide metadata, so Dask is running the"
```
```suggestion
            "\nYou did not provide metadata for the output, so Dask is running the"
```

I personally got slightly confused about which metadata was being talked about here: `input_dask_ar._meta` or metadata passed in with `predict_meta`/`predict_proba_meta`/`transform_meta`. I think this will only be confusing for developers who have read the code above this bit, since users will just see the message. In any case, it might help to make the message a bit clearer.

VibhuJawa replied:
Reworded it to make as clear as i could. 

This is how it will look the user now. 

```python
UserWarning: Metadata for predict_proba is not provided, so Dask is running the predict_proba function on a small dataset to guess output metadata. As a result, It is possible that Dask will guess incorrectly.
To silence this warning, provide explicit `predict_proba_meta` to the dask_ml.wrapper.
Example: 
wrap_clf = dask_ml.wrappers.Incremental(GradientBoostingClassifier(), predict_proba_meta = np.array([1],dtype=np.int8))
  warnings.warn(msg)
```

Thanks for the inputs, please  feel free to let me know if it is still confusing.

GenevieveBuckley commented on dask_ml/wrappers.py:
```diff
@@ -662,9 +663,28 @@ def _get_output_dask_ar_meta_for_estimator(model_fn, estimator, input_dask_ar):
+            "\nYou did not provide metadata, so Dask is running the"
+            "function on a small dataset to guess output types. "
+            "It is possible that Dask will guess incorrectly.\n"
+            "To provide an explicit output types or to silence this message, "
```
```suggestion
            "To provide an explicit output metadata to silence this message, "
```

I don't think "or" belongs in this sentence? Providing the explicit output metadata is the thing that silences this message, it's not two separate things, isn't that right? Also in this sentence it seems like `output types` => `output metadata`. 

If they are in fact two separate things, you'll need instructions on how to do both (eg: kwargs for providing explicit output type, as well as the kwargs for providing the output metadata)

VibhuJawa replied:
Addressed this with commit https://github.com/dask/dask-ml/pull/889/commits/48645448147126c6a343d6d085dd571be0cd1665

GenevieveBuckley reviewed (commented):
I think the general approach here seems good. I've left a few comments, mostly about minor things.

VibhuJawa commented:
Thanks a lot for reviewing the PR @GenevieveBuckley and providing actionable inputs.  I  have addressed your reviews and completed my todo lists. This PR is now ready for another round of reviews.

GenevieveBuckley commented:
Thanks @VibhuJawa! Someone else from the maintenance team will have a look at this PR next.

TomAugspurger commented:
Thanks @VibhuJawa, and thanks for the review @GenevieveBuckley!

I'll make a release with this now.
